lib/rex/post/meterpreter/extensions/stdapi/sys/config.rb

```diff
@@ -90,7 +90,7 @@ def getenvs(*var_names)
     response.each(TLV_TYPE_ENV_GROUP) do |env|
       var_name = env.get_tlv_value(TLV_TYPE_ENV_VARIABLE)
       var_value = env.get_tlv_value(TLV_TYPE_ENV_VALUE)
-      result[var_name] = var_value
+      result[var_name] = var_value.scan(/[[:print:]]/).join
```

Comment:
> when the environment variable was not present, the get_tlvs() function present in lib/rex/post/meterpreter/packet.rb returned non printable ascii values.

This solution looks like a work around to me 👀

Does this require a change in Meterpreter? i.e. why does Meterpreter send back junk values when the env isn't present?

Reply:
The get_tlv() function when invoked does give out random hex values when no env is present. That happens in nearly one out of 10 fetches. My assumption is that something weird is happening on the windows side when a request is made for the  non existent env vars. Even I was uncomfortable with this solution. But since the get_tlv is polymorphic and is used in many places throughout the meterpreter codebase. Felt too risky modifying it. Thats why the change is only made at the assignment level rather than in get_tlv function

Reply:
I think the Meterpreter code needs to be updated, rather than this ruby code.


It looks like the `GetEnvironmentVariableW` method is called, but the result value isn't checked to see if the return value is 0:

https://github.com/rapid7/metasploit-payloads/blob/6e08d1f9812aa4d7a76b451fd5e0bae03975bb91/c/meterpreter/source/extensions/stdapi/server/sys/config/config.c#L88-L98

From the [docs](https://learn.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-getenvironmentvariablew):

> If the function succeeds, the return value is the number of characters stored in the buffer pointed to by lpBuffer, not including the terminating null character.
> ...
> If the function fails, the return value is zero. If the specified environment variable was not found in the environment block, [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror) returns ERROR_ENVVAR_NOT_FOUND.

And it looks like the logic for copying the env var value relies on strlen, rather than the return value of `GetEnvironmentVariableW` - which is why we end up getting unexpected bytes back from the C Meterpreter

https://github.com/rapid7/metasploit-payloads/blob/6e08d1f9812aa4d7a76b451fd5e0bae03975bb91/c/meterpreter/source/extensions/stdapi/server/sys/config/config.c#L26-L28

Reply:
With that knowledge, I think there should be an update to check the result of `GetEnvironmentVariableW` - and if the return value is 0 - we should not return the env pair to the user

Reply:
Understood, Thanks @adfoster-r7 for the suggestion. creating another pull request with the suggested fix.